platform/logger.cpp

```diff
@@ -0,0 +1,240 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) && !defined(NDEBUG)
+#include <string.h>
+#include "platform/logger_internal.h"
+#include "hal/ticker_api.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_interface.h"
+#include "mbed_events.h"
+
+using namespace mbed;
+
+#if defined (MBED_ID_BASED_TRACING)
+static LOG_DATA_TYPE_ log_id_buf_[LOG_SINGLE_STR_SIZE_];
+static uint32_t log_count_ = 0;
+#endif
+
+// Globals related to ISR logging
+static LOG_DATA_TYPE_ log_isr_str_[LOG_SINGLE_STR_SIZE_];
+static EventQueue *log_isr_equeue_;
+
+// Globals related to time printing
+static const ticker_data_t *const log_ticker_ = get_us_ticker_data();
+static bool time_enable_ = true;
+
+// Globals related to external buffer capturing
+static LOG_DATA_TYPE_ *extern_buf_ = NULL;
+static uint32_t xbuf_count_ = 0;
+
+static void log_update_buf(const char *data, int32_t size)
+{
+    if ((NULL == extern_buf_) || (NULL == data) || (size <= 0)) {
+        return;
+    }
+    uint32_t count = 0;
+    while(size--)
+    {
+        extern_buf_[xbuf_count_++] = data[count++];
+        xbuf_count_ %= MBED_CONF_EXTERNAL_BUFFER_SIZE;
+    }
+}
+
+#if defined (MBED_ID_BASED_TRACING)
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+static void log_isr_id_data()
+{
+    mbed_log_lock();
+    uint32_t count = 0;
+    while(log_count_--) {
+        fprintf(stderr, "0x%x ", log_isr_str_[count++]);
+    }
+    mbed_log_unlock();
+}
+#else
+// Note: ISR data is lossy, and with too many interrupts it might be
+// over-written and last valid ISR data will be printed.
+static void log_isr_queue()
+{
+    mbed_log_lock();
+    fputs(log_isr_str_, stderr);
+    mbed_log_unlock();
+}
+
+static void log_str_isr_data(const char *format, va_list args)
+{
+    core_util_critical_section_enter();
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    int32_t count = 0;
+    if (time_enable_) {
+        count = snprintf(log_isr_str_, LOG_SINGLE_STR_SIZE_, "[%-8lld]", time);
+    }
+    count = vsnprintf(log_isr_str_+count, (LOG_SINGLE_STR_SIZE_-count), format, args);
+    if (NULL != extern_buf_) {
+        log_update_buf(log_isr_str_, count);
+    } else {
+        log_isr_equeue_->call(log_isr_queue);
+    }
+    core_util_critical_section_exit();
+}
+
+static void log_str_usr_data(const char *format, va_list args)
+{
+    mbed_log_lock();
+    LOG_DATA_TYPE_ one_line[LOG_SINGLE_STR_SIZE_ << 1];
+    uint32_t size = (LOG_SINGLE_STR_SIZE_ << 1);
+
+    volatile uint64_t time = ticker_read_us(log_ticker_);
+    int32_t count = 0;
+    if (time_enable_) {
+        count = snprintf(one_line, size, "[%-8lld]", time);
+    }
+    count = vsnprintf(one_line+count, (size-count), format, args);
```

Comment:
same question, generally, how error return values (negative) from xxnprintf () are handled so that they are not causing unexpected behaviour (mostly corner cases)?